// @deepgram/sdk/dist/esm/api/resources/listen/resources/v1/client/Client.mjs
// This file was auto-generated by Fern from our API Definition.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { normalizeClientOptionsWithAuth } from "../../../../../../BaseClient.mjs";
import { mergeHeaders, mergeOnlyDefinedHeaders } from "../../../../../../core/headers.mjs";
import * as core from "../../../../../../core/index.mjs";
import { toJson } from "../../../../../../core/json.mjs";
import * as environments from "../../../../../../environments.mjs";
import { MediaClient } from "../resources/media/client/Client.mjs";
import { V1Socket } from "./Socket.mjs";
export class V1Client {
    constructor(options = {}) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    get media() {
        var _a;
        return ((_a = this._media) !== null && _a !== void 0 ? _a : (this._media = new MediaClient(this._options)));
    }
    connect(args) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const { callback, callback_method: callbackMethod, channels, detect_entities: detectEntities, diarize, dictation, encoding, endpointing, extra, interim_results: interimResults, keyterm, keywords, language, mip_opt_out: mipOptOut, model, multichannel, numerals, profanity_filter: profanityFilter, punctuate, redact, replace, sample_rate: sampleRate, search, smart_format: smartFormat, tag, utterance_end_ms: utteranceEndMs, vad_events: vadEvents, version, queryParams, headers, debug, reconnectAttempts, connectionTimeoutInSeconds, abortSignal, } = args;
            const _queryParams = {
                callback: callback != null ? (typeof callback === "string" ? callback : toJson(callback)) : undefined,
                callback_method: callbackMethod != null ? callbackMethod : undefined,
                channels: channels != null ? (typeof channels === "string" ? channels : toJson(channels)) : undefined,
                detect_entities: detectEntities != null ? detectEntities : undefined,
                diarize: diarize != null ? diarize : undefined,
                dictation: dictation != null ? dictation : undefined,
                encoding: encoding != null ? encoding : undefined,
                endpointing: endpointing != null ? (typeof endpointing === "string" ? endpointing : toJson(endpointing)) : undefined,
                extra: extra != null ? (typeof extra === "string" ? extra : toJson(extra)) : undefined,
                interim_results: interimResults != null ? interimResults : undefined,
                keyterm: keyterm != null ? (typeof keyterm === "string" ? keyterm : toJson(keyterm)) : undefined,
                keywords: keywords != null ? (typeof keywords === "string" ? keywords : toJson(keywords)) : undefined,
                language: language != null ? (typeof language === "string" ? language : toJson(language)) : undefined,
                mip_opt_out: mipOptOut != null ? (typeof mipOptOut === "string" ? mipOptOut : toJson(mipOptOut)) : undefined,
                model,
                multichannel: multichannel != null ? multichannel : undefined,
                numerals: numerals != null ? numerals : undefined,
                profanity_filter: profanityFilter != null ? profanityFilter : undefined,
                punctuate: punctuate != null ? punctuate : undefined,
                redact: redact != null ? redact : undefined,
                replace: replace != null ? (typeof replace === "string" ? replace : toJson(replace)) : undefined,
                sample_rate: sampleRate != null ? (typeof sampleRate === "string" ? sampleRate : toJson(sampleRate)) : undefined,
                search: search != null ? (typeof search === "string" ? search : toJson(search)) : undefined,
                smart_format: smartFormat != null ? smartFormat : undefined,
                tag: tag != null ? (typeof tag === "string" ? tag : toJson(tag)) : undefined,
                utterance_end_ms: utteranceEndMs != null
                    ? typeof utteranceEndMs === "string"
                        ? utteranceEndMs
                        : toJson(utteranceEndMs)
                    : undefined,
                vad_events: vadEvents != null ? vadEvents : undefined,
                version: version != null ? (typeof version === "string" ? version : toJson(version)) : undefined,
            };
            const _headers = mergeHeaders(mergeOnlyDefinedHeaders({ Authorization: args.Authorization }), headers);
            const socket = new core.ReconnectingWebSocket({
                url: core.url.join((_a = (yield core.Supplier.get(this._options.baseUrl))) !== null && _a !== void 0 ? _a : ((_b = (yield core.Supplier.get(this._options.environment))) !== null && _b !== void 0 ? _b : environments.DeepgramEnvironment.Production).production, "/v1/listen"),
                protocols: [],
                queryParameters: Object.assign(Object.assign({}, _queryParams), queryParams),
                headers: _headers,
                options: {
                    debug: debug !== null && debug !== void 0 ? debug : false,
                    maxRetries: reconnectAttempts !== null && reconnectAttempts !== void 0 ? reconnectAttempts : 30,
                    connectionTimeout: connectionTimeoutInSeconds != null ? connectionTimeoutInSeconds * 1000 : undefined,
                },
                abortSignal: abortSignal,
            });
            return new V1Socket({ socket });
        });
    }
}
