// @deepgram/sdk/dist/esm/Client.mjs
// This file was auto-generated by Fern from our API Definition.
import { AgentClient } from "./api/resources/agent/client/Client.mjs";
import { AuthClient } from "./api/resources/auth/client/Client.mjs";
import { ListenClient } from "./api/resources/listen/client/Client.mjs";
import { ManageClient } from "./api/resources/manage/client/Client.mjs";
import { ReadClient } from "./api/resources/read/client/Client.mjs";
import { SelfHostedClient } from "./api/resources/selfHosted/client/Client.mjs";
import { SpeakClient } from "./api/resources/speak/client/Client.mjs";
import { normalizeClientOptionsWithAuth } from "./BaseClient.mjs";
export class DeepgramClient {
    constructor(options = {}) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    get agent() {
        var _a;
        return ((_a = this._agent) !== null && _a !== void 0 ? _a : (this._agent = new AgentClient(this._options)));
    }
    get auth() {
        var _a;
        return ((_a = this._auth) !== null && _a !== void 0 ? _a : (this._auth = new AuthClient(this._options)));
    }
    get listen() {
        var _a;
        return ((_a = this._listen) !== null && _a !== void 0 ? _a : (this._listen = new ListenClient(this._options)));
    }
    get manage() {
        var _a;
        return ((_a = this._manage) !== null && _a !== void 0 ? _a : (this._manage = new ManageClient(this._options)));
    }
    get read() {
        var _a;
        return ((_a = this._read) !== null && _a !== void 0 ? _a : (this._read = new ReadClient(this._options)));
    }
    get selfHosted() {
        var _a;
        return ((_a = this._selfHosted) !== null && _a !== void 0 ? _a : (this._selfHosted = new SelfHostedClient(this._options)));
    }
    get speak() {
        var _a;
        return ((_a = this._speak) !== null && _a !== void 0 ? _a : (this._speak = new SpeakClient(this._options)));
    }
}
