// @deepgram/sdk/dist/esm/CustomClient.mjs
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { DeepgramClient } from "./Client.mjs";
import { ReconnectingWebSocket } from "./core/websocket/ws.mjs";
import { AgentClient as AgentClientImpl } from "./api/resources/agent/client/Client.mjs";
import { ListenClient as ListenClientImpl } from "./api/resources/listen/client/Client.mjs";
import { SpeakClient as SpeakClientImpl } from "./api/resources/speak/client/Client.mjs";
import { V1Client as AgentV1Client } from "./api/resources/agent/resources/v1/client/Client.mjs";
import { V1Client as ListenV1Client } from "./api/resources/listen/resources/v1/client/Client.mjs";
import { V2Client as ListenV2Client } from "./api/resources/listen/resources/v2/client/Client.mjs";
import { V1Client as SpeakV1Client } from "./api/resources/speak/resources/v1/client/Client.mjs";
import { V1Socket as AgentV1Socket } from "./api/resources/agent/resources/v1/client/Socket.mjs";
import { V1Socket as ListenV1Socket } from "./api/resources/listen/resources/v1/client/Socket.mjs";
import { V2Socket as ListenV2Socket } from "./api/resources/listen/resources/v2/client/Socket.mjs";
import { V1Socket as SpeakV1Socket } from "./api/resources/speak/resources/v1/client/Socket.mjs";
import { mergeHeaders } from "./core/headers.mjs";
import { fromJson } from "./core/json.mjs";
import * as core from "./core/index.mjs";
import * as environments from "./environments.mjs";
import { RUNTIME } from "./core/runtime/index.mjs";
// Default WebSocket connection timeout in milliseconds
const DEFAULT_CONNECTION_TIMEOUT_MS = 10000;
// Keys present in every ConnectArgs interface that control the WebSocket connection itself.
// Every other key in ConnectArgs is treated as an API query parameter.
const WEBSOCKET_OPTION_KEYS = new Set([
    "Authorization",
    "headers",
    "debug",
    "reconnectAttempts",
    "connectionTimeoutInSeconds",
    "abortSignal",
    "queryParams",
]);
// ws for Node.js - loaded lazily to support CJS, ESM, and browser builds.
// A static import of "module" (for createRequire) would break the browser bundle,
// so we detect the environment at runtime and use an opaque dynamic import in ESM
// Node so bundlers cannot statically analyse and reject it.
let NodeWebSocket;
let _wsInitialized = false;
function loadNodeWebSocket() {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        if (_wsInitialized)
            return;
        _wsInitialized = true;
        try {
            if (typeof require !== "undefined") {
                // CJS: require is injected as a module-scoped binding
                // eslint-disable-next-line @typescript-eslint/no-require-imports
                let ws = require("ws");
                NodeWebSocket = ws.WebSocket || ws.default || ws;
            }
            else if (typeof process !== "undefined" && ((_a = process.versions) === null || _a === void 0 ? void 0 : _a.node)) {
                // ESM Node: require is not defined. Wrap import() in new Function so
                // bundlers (esbuild, webpack) cannot statically resolve or reject "ws".
                // eslint-disable-next-line no-new-func
                const dynamicImport = new Function("specifier", "return import(specifier)");
                const ws = yield dynamicImport("ws");
                NodeWebSocket = ws.WebSocket || ws.default || ws;
            }
            // Browser: process.versions?.node is undefined → NodeWebSocket stays undefined
        }
        catch (_b) {
            // ws not available or failed to load
            NodeWebSocket = undefined;
        }
    });
}
// Helper function to generate UUID that works in both Node.js and browser
function generateUUID() {
    // Try global crypto.randomUUID first (works in both Node.js 14.18+ and modern browsers)
    if (typeof crypto !== "undefined" && typeof crypto.randomUUID === "function") {
        return crypto.randomUUID();
    }
    // Fallback for Node.js: use the crypto module
    if (RUNTIME.type === "node") {
        try {
            // eslint-disable-next-line @typescript-eslint/no-require-imports
            const nodeCrypto = require("crypto");
            return nodeCrypto.randomUUID();
        }
        catch (_a) {
            // Fallback if crypto module is not available
        }
    }
    // Final fallback: manual UUID generation (RFC4122 version 4)
    // This should work everywhere but is less secure than crypto.randomUUID()
    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
        const r = (Math.random() * 16) | 0;
        const v = c === "x" ? r : (r & 0x3) | 0x8;
        return v.toString(16);
    });
}
/**
 * Wrapper auth provider that adds "Token " prefix to API keys.
 * The auto-generated HeaderAuthProvider doesn't add the prefix, so we wrap it here.
 */
class ApiKeyAuthProviderWrapper {
    constructor(originalProvider) {
        this.originalProvider = originalProvider;
    }
    getAuthRequest(arg) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const authRequest = yield this.originalProvider.getAuthRequest(arg);
            const authHeader = ((_a = authRequest.headers) === null || _a === void 0 ? void 0 : _a.Authorization) || ((_b = authRequest.headers) === null || _b === void 0 ? void 0 : _b.authorization);
            // If the header doesn't already have a scheme prefix, add "Token " prefix for API keys
            if (authHeader && typeof authHeader === "string") {
                // Only add prefix if it doesn't already have Bearer or Token prefix
                if (!authHeader.startsWith("Bearer ") && !authHeader.startsWith("Token ") && !authHeader.startsWith("token ")) {
                    return {
                        headers: Object.assign(Object.assign({}, authRequest.headers), { Authorization: `Token ${authHeader}` }),
                    };
                }
            }
            return authRequest;
        });
    }
}
/**
 * Wrapper auth provider that checks for accessToken first (Bearer scheme)
 * before falling back to the original auth provider (Token scheme for API keys).
 */
class AccessTokenAuthProviderWrapper {
    constructor(originalProvider, accessToken) {
        this.originalProvider = originalProvider;
        this.accessToken = accessToken;
    }
    getAuthRequest(arg) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            // Check for access token first (highest priority)
            // Access tokens use Bearer scheme, API keys use Token scheme
            const accessToken = (_a = (yield core.Supplier.get(this.accessToken))) !== null && _a !== void 0 ? _a : (_b = process.env) === null || _b === void 0 ? void 0 : _b.DEEPGRAM_ACCESS_TOKEN;
            if (accessToken != null) {
                return {
                    headers: { Authorization: `Bearer ${accessToken}` },
                };
            }
            // Fall back to original provider (which handles API keys)
            return this.originalProvider.getAuthRequest(arg);
        });
    }
}
/**
 * Custom wrapper around DeepgramClient that ensures the custom websocket implementation
 * from ws.ts is always used, even if the auto-generated code changes.
 */
export class CustomDeepgramClient extends DeepgramClient {
    constructor(options = {}) {
        // Generate a UUID for the session ID
        const sessionId = generateUUID();
        // Add the session ID to headers so it's included in all REST requests
        const optionsWithSessionId = Object.assign(Object.assign({}, options), { headers: Object.assign(Object.assign({}, options.headers), { "x-deepgram-session-id": sessionId }) });
        super(optionsWithSessionId);
        this._sessionId = sessionId;
        // Always wrap the auth provider to add "Token " prefix to API keys
        // The auto-generated HeaderAuthProvider doesn't add the prefix
        this._options.authProvider = new ApiKeyAuthProviderWrapper(this._options.authProvider);
        // Wrap again to handle accessToken if provided
        // This ensures accessToken takes priority over apiKey/env var
        if (options.accessToken != null) {
            this._options.authProvider = new AccessTokenAuthProviderWrapper(this._options.authProvider, options.accessToken);
        }
    }
    /**
     * Get the session ID that was generated for this client instance.
     */
    get sessionId() {
        return this._sessionId;
    }
    /**
     * Override the agent getter to return a wrapped client that ensures
     * the custom websocket implementation is used.
     */
    get agent() {
        if (!this._customAgent) {
            // Create a wrapper that ensures custom websocket is used
            this._customAgent = new WrappedAgentClient(this._options);
        }
        return this._customAgent;
    }
    /**
     * Override the listen getter to return a wrapped client that ensures
     * the custom websocket implementation is used.
     */
    get listen() {
        if (!this._customListen) {
            // Create a wrapper that ensures custom websocket is used
            this._customListen = new WrappedListenClient(this._options);
        }
        return this._customListen;
    }
    /**
     * Override the speak getter to return a wrapped client that ensures
     * the custom websocket implementation is used.
     */
    get speak() {
        if (!this._customSpeak) {
            // Create a wrapper that ensures custom websocket is used
            this._customSpeak = new WrappedSpeakClient(this._options);
        }
        return this._customSpeak;
    }
}
/**
 * Wrapper for AgentClient that ensures custom websocket implementation is used.
 *
 * This wrapper exists to guarantee that our custom WebSocket implementation
 * (from ws.ts) continues to be used even after the SDK code is auto-generated
 * by Fern. Without this wrapper, Fern regeneration could overwrite the client
 * with a different WebSocket implementation.
 */
class WrappedAgentClient extends AgentClientImpl {
    get v1() {
        return new WrappedAgentV1Client(this._options);
    }
}
/**
 * Wrapper for ListenClient that ensures custom websocket implementation is used.
 *
 * This wrapper exists to guarantee that our custom WebSocket implementation
 * (from ws.ts) continues to be used even after the SDK code is auto-generated
 * by Fern. Without this wrapper, Fern regeneration could overwrite the client
 * with a different WebSocket implementation.
 */
class WrappedListenClient extends ListenClientImpl {
    get v1() {
        return new WrappedListenV1Client(this._options);
    }
    get v2() {
        return new WrappedListenV2Client(this._options);
    }
}
/**
 * Wrapper for SpeakClient that ensures custom websocket implementation is used.
 *
 * This wrapper exists to guarantee that our custom WebSocket implementation
 * (from ws.ts) continues to be used even after the SDK code is auto-generated
 * by Fern. Without this wrapper, Fern regeneration could overwrite the client
 * with a different WebSocket implementation.
 */
class WrappedSpeakClient extends SpeakClientImpl {
    get v1() {
        return new WrappedSpeakV1Client(this._options);
    }
}
/**
 * Helper function to resolve Suppliers in headers to their actual values.
 */
function resolveHeaders(headers) {
    return __awaiter(this, void 0, void 0, function* () {
        const resolved = {};
        for (const [key, value] of Object.entries(headers)) {
            if (value == null) {
                continue;
            }
            // Resolve Supplier if it's a Supplier, otherwise use the value directly
            const resolvedValue = yield core.Supplier.get(value);
            if (resolvedValue != null) {
                resolved[key] = resolvedValue;
            }
        }
        return resolved;
    });
}
/**
 * Builds API query parameters from a ConnectArgs object, excluding all WebSocket
 * infrastructure keys. This means any new typed parameter added to ConnectArgs by
 * the generator is automatically included without needing manual updates here.
 * An explicit `queryParams` override is merged in last.
 */
function buildQueryParams(args) {
    const result = {};
    for (const [key, value] of Object.entries(args)) {
        if (!WEBSOCKET_OPTION_KEYS.has(key) && value != null) {
            result[key] = value;
        }
    }
    if (args.queryParams != null && typeof args.queryParams === "object") {
        Object.assign(result, args.queryParams);
    }
    return result;
}
/**
 * Helper function to get WebSocket class and handle headers/protocols based on runtime.
 * In Node.js, use the 'ws' library which supports headers.
 * In browser, use Sec-WebSocket-Protocol for authentication since headers aren't supported.
 */
function getWebSocketOptions(headers) {
    const options = {};
    // Check if we're in a browser environment (browser or web-worker)
    const isBrowser = RUNTIME.type === "browser" || RUNTIME.type === "web-worker";
    // Extract session ID header
    const sessionIdHeader = headers["x-deepgram-session-id"] || headers["X-Deepgram-Session-Id"];
    // In Node.js, ensure we use the 'ws' library which supports headers
    if (RUNTIME.type === "node" && NodeWebSocket) {
        options.WebSocket = NodeWebSocket;
        options.headers = headers;
    }
    else if (isBrowser) {
        // In browser, native WebSocket doesn't support custom headers
        // Extract Authorization header and use Sec-WebSocket-Protocol instead
        const authHeader = headers.Authorization || headers.authorization;
        const browserHeaders = Object.assign({}, headers);
        // Remove Authorization and session ID from headers since they won't work in browser
        delete browserHeaders.Authorization;
        delete browserHeaders.authorization;
        delete browserHeaders["x-deepgram-session-id"];
        delete browserHeaders["X-Deepgram-Session-Id"];
        options.headers = browserHeaders;
        // Build protocols array for browser WebSocket
        const protocols = [];
        // If we have an Authorization header, extract the token and format as protocols
        // Deepgram expects:
        // - For API keys: Sec-WebSocket-Protocol: token,API_KEY_GOES_HERE
        // - For Bearer tokens: Sec-WebSocket-Protocol: bearer,TOKEN_GOES_HERE
        // The comma separates multiple protocols, so we pass them as an array
        if (authHeader && typeof authHeader === "string") {
            if (authHeader.startsWith("Token ")) {
                // API key: "Token API_KEY" -> ["token", "API_KEY"]
                const apiKey = authHeader.substring(6); // Remove "Token " prefix
                protocols.push("token", apiKey);
            }
            else if (authHeader.startsWith("Bearer ")) {
                // Access token: "Bearer TOKEN" -> ["bearer", "TOKEN"]
                const token = authHeader.substring(7); // Remove "Bearer " prefix
                protocols.push("bearer", token);
            }
            else {
                // Fallback: use the entire header value if it doesn't match expected format
                protocols.push(authHeader);
            }
        }
        // Add session ID as a protocol for browser WebSocket
        if (sessionIdHeader && typeof sessionIdHeader === "string") {
            protocols.push("x-deepgram-session-id", sessionIdHeader);
        }
        if (protocols.length > 0) {
            options.protocols = protocols;
        }
    }
    else {
        // Fallback for other environments
        options.headers = headers;
    }
    return options;
}
/**
 * Helper function to setup binary-aware message handling for WebSocket sockets.
 * Handles both text (JSON) and binary messages correctly.
 */
function setupBinaryHandling(socket, eventHandlers) {
    var _a;
    const binaryAwareHandler = (event) => {
        var _a, _b, _c;
        // Handle both text (JSON) and binary messages
        if (typeof event.data === "string") {
            try {
                const data = fromJson(event.data);
                (_a = eventHandlers.message) === null || _a === void 0 ? void 0 : _a.call(eventHandlers, data);
            }
            catch (error) {
                // If JSON parsing fails, pass the raw string
                (_b = eventHandlers.message) === null || _b === void 0 ? void 0 : _b.call(eventHandlers, event.data);
            }
        }
        else {
            // Binary data - pass through as-is
            (_c = eventHandlers.message) === null || _c === void 0 ? void 0 : _c.call(eventHandlers, event.data);
        }
    };
    // Remove the original handler and add our binary-aware one
    const socketAny = socket;
    if ((_a = socketAny._listeners) === null || _a === void 0 ? void 0 : _a.message) {
        // Remove all message listeners
        socketAny._listeners.message.forEach((listener) => {
            socket.removeEventListener("message", listener);
        });
    }
    // Add our binary-aware handler
    socket.addEventListener("message", binaryAwareHandler);
    return binaryAwareHandler;
}
/**
 * Helper to prevent duplicate event listeners on WebSocket connections.
 * This removes all event listeners that were registered by the auto-generated
 * Socket class constructor, preventing duplicate event firing when connect() is called.
 */
function preventDuplicateEventListeners(socket, handlers) {
    // Remove the handlers that were added by the auto-generated constructor
    if (handlers.handleOpen) {
        socket.removeEventListener("open", handlers.handleOpen);
    }
    if (handlers.handleMessage) {
        socket.removeEventListener("message", handlers.handleMessage);
    }
    if (handlers.handleClose) {
        socket.removeEventListener("close", handlers.handleClose);
    }
    if (handlers.handleError) {
        socket.removeEventListener("error", handlers.handleError);
    }
}
/**
 * Helper function to reset socket connection state before connecting.
 * Ensures _connectLock is reset if the socket is in CLOSED state.
 */
function resetSocketConnectionState(socket) {
    if (socket.readyState === socket.CLOSED) {
        // Force a fresh reconnect to ensure _connectLock is reset
        socket._connectLock = false;
        socket._shouldReconnect = true;
    }
}
/**
 * Helper function to create a WebSocket connection with common setup logic.
 * Handles authentication, header merging, and WebSocket configuration.
 * This reduces duplication across all Wrapped*Client classes.
 */
function createWebSocketConnection(_a) {
    return __awaiter(this, arguments, void 0, function* ({ options, urlPath, environmentKey, queryParams, headers, debug, reconnectAttempts, connectionTimeoutInSeconds, abortSignal, }) {
        var _b, _c, _d, _e, _f, _g;
        // Ensure ws is loaded for Node.js environments (no-op after first call)
        yield loadNodeWebSocket();
        // Get Authorization from authProvider (cast to any to access internal property)
        const authRequest = yield ((_b = options.authProvider) === null || _b === void 0 ? void 0 : _b.getAuthRequest());
        // Merge headers from options (which includes session ID), auth headers, and request headers
        const mergedHeaders = mergeHeaders((_c = options.headers) !== null && _c !== void 0 ? _c : {}, (_d = authRequest === null || authRequest === void 0 ? void 0 : authRequest.headers) !== null && _d !== void 0 ? _d : {}, headers);
        // Resolve any Suppliers in headers to actual values
        const _headers = yield resolveHeaders(mergedHeaders);
        // Get WebSocket options with proper header handling
        const wsOptions = getWebSocketOptions(_headers);
        // Get the appropriate base URL for the environment
        const baseUrl = (_e = (yield core.Supplier.get(options.baseUrl))) !== null && _e !== void 0 ? _e : ((_f = (yield core.Supplier.get(options.environment))) !== null && _f !== void 0 ? _f : environments.DeepgramEnvironment.Production)[environmentKey];
        // Create and return the ReconnectingWebSocket
        return new ReconnectingWebSocket({
            url: core.url.join(baseUrl, urlPath),
            protocols: (_g = wsOptions.protocols) !== null && _g !== void 0 ? _g : [],
            queryParameters: queryParams,
            headers: wsOptions.headers,
            options: {
                WebSocket: wsOptions.WebSocket,
                debug: debug !== null && debug !== void 0 ? debug : false,
                maxRetries: reconnectAttempts !== null && reconnectAttempts !== void 0 ? reconnectAttempts : 30,
                startClosed: true,
                connectionTimeout: connectionTimeoutInSeconds != null ? connectionTimeoutInSeconds * 1000 : DEFAULT_CONNECTION_TIMEOUT_MS,
            },
            abortSignal,
        });
    });
}
/**
 * Wrapper for Agent V1Client that overrides connect to use custom websocket.
 *
 * This wrapper ensures that the connect() method uses our custom ReconnectingWebSocket
 * implementation instead of any auto-generated WebSocket handling. This guarantees
 * consistent behavior across Fern regenerations and allows us to customize
 * connection setup, authentication, and header handling.
 */
class WrappedAgentV1Client extends AgentV1Client {
    connect() {
        return __awaiter(this, arguments, void 0, function* (args = {}) {
            const { headers, debug, reconnectAttempts, connectionTimeoutInSeconds, abortSignal } = args;
            const socket = yield createWebSocketConnection({
                options: this._options,
                urlPath: "/v1/agent/converse",
                environmentKey: 'agent',
                queryParams: buildQueryParams(args),
                headers,
                debug,
                reconnectAttempts,
                connectionTimeoutInSeconds,
                abortSignal,
            });
            return new WrappedAgentV1Socket({ socket });
        });
    }
    /**
     * Creates a WebSocket connection object without actually connecting.
     * This is an alias for connect() with clearer naming - the returned socket
     * is not connected until you call socket.connect().
     *
     * Usage:
     * ```typescript
     * const socket = await client.agent.v1.createConnection();
     * socket.on('open', () => console.log('Connected!'));
     * socket.on('message', (msg) => console.log('Message:', msg));
     * socket.connect(); // Actually initiates the connection
     * ```
     */
    createConnection() {
        return __awaiter(this, arguments, void 0, function* (args = {}) {
            return this.connect(args);
        });
    }
}
/**
 * Wrapper for Agent V1Socket that handles binary messages correctly.
 *
 * This wrapper ensures that both text (JSON) and binary WebSocket messages are
 * handled properly. The auto-generated socket class may not handle binary data
 * correctly, so we override the message handling to support both formats.
 * This is critical for audio and other binary data streaming.
 */
class WrappedAgentV1Socket extends AgentV1Socket {
    constructor(args) {
        super(args);
        this.setupBinaryHandling();
    }
    setupBinaryHandling() {
        this.binaryAwareHandler = setupBinaryHandling(this.socket, this.eventHandlers);
    }
    connect() {
        // Remove duplicate listeners before calling super.connect()
        const socketAny = this;
        preventDuplicateEventListeners(this.socket, {
            handleOpen: socketAny.handleOpen,
            handleMessage: socketAny.handleMessage,
            handleClose: socketAny.handleClose,
            handleError: socketAny.handleError
        });
        resetSocketConnectionState(this.socket);
        super.connect();
        // Re-setup binary handling in case connect() re-registered handlers
        this.setupBinaryHandling();
        return this;
    }
}
/**
 * Wrapper for Listen V1Client that overrides connect to use custom websocket.
 *
 * This wrapper ensures that the connect() method uses our custom ReconnectingWebSocket
 * implementation instead of any auto-generated WebSocket handling. This guarantees
 * consistent behavior across Fern regenerations and allows us to customize
 * connection setup, authentication, and header handling.
 */
class WrappedListenV1Client extends ListenV1Client {
    connect(args) {
        return __awaiter(this, void 0, void 0, function* () {
            const { headers, debug, reconnectAttempts, connectionTimeoutInSeconds, abortSignal } = args;
            const socket = yield createWebSocketConnection({
                options: this._options,
                urlPath: "/v1/listen",
                environmentKey: 'production',
                queryParams: buildQueryParams(args),
                headers,
                debug,
                reconnectAttempts,
                connectionTimeoutInSeconds,
                abortSignal,
            });
            return new WrappedListenV1Socket({ socket });
        });
    }
    /**
     * Creates a WebSocket connection object without actually connecting.
     * This is an alias for connect() with clearer naming - the returned socket
     * is not connected until you call socket.connect().
     *
     * Usage:
     * ```typescript
     * const socket = await client.listen.v1.createConnection({ model: 'nova-3' });
     * socket.on('open', () => console.log('Connected!'));
     * socket.on('message', (msg) => console.log('Transcript:', msg));
     * socket.connect(); // Actually initiates the connection
     * ```
     */
    createConnection(args) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.connect(args);
        });
    }
}
/**
 * Wrapper for Listen V1Socket that handles binary messages correctly.
 *
 * This wrapper ensures that both text (JSON) and binary WebSocket messages are
 * handled properly. The auto-generated socket class may not handle binary data
 * correctly, so we override the message handling to support both formats.
 * This is critical for audio and other binary data streaming.
 */
class WrappedListenV1Socket extends ListenV1Socket {
    constructor(args) {
        super(args);
        this.setupBinaryHandling();
    }
    setupBinaryHandling() {
        this.binaryAwareHandler = setupBinaryHandling(this.socket, this.eventHandlers);
    }
    connect() {
        // Remove duplicate listeners before calling super.connect()
        const socketAny = this;
        preventDuplicateEventListeners(this.socket, {
            handleOpen: socketAny.handleOpen,
            handleMessage: socketAny.handleMessage,
            handleClose: socketAny.handleClose,
            handleError: socketAny.handleError
        });
        resetSocketConnectionState(this.socket);
        super.connect();
        this.setupBinaryHandling();
        return this;
    }
}
/**
 * Wrapper for Listen V2Client that overrides connect to use custom websocket.
 *
 * This wrapper ensures that the connect() method uses our custom ReconnectingWebSocket
 * implementation instead of any auto-generated WebSocket handling. This guarantees
 * consistent behavior across Fern regenerations and allows us to customize
 * connection setup, authentication, and header handling.
 */
class WrappedListenV2Client extends ListenV2Client {
    connect(args) {
        return __awaiter(this, void 0, void 0, function* () {
            const { headers, debug, reconnectAttempts, connectionTimeoutInSeconds, abortSignal } = args;
            const socket = yield createWebSocketConnection({
                options: this._options,
                urlPath: "/v2/listen",
                environmentKey: 'production',
                queryParams: buildQueryParams(args),
                headers,
                debug,
                reconnectAttempts,
                connectionTimeoutInSeconds,
                abortSignal,
            });
            return new WrappedListenV2Socket({ socket });
        });
    }
    /**
     * Creates a WebSocket connection object without actually connecting.
     * This is an alias for connect() with clearer naming - the returned socket
     * is not connected until you call socket.connect().
     *
     * Usage:
     * ```typescript
     * const socket = await client.listen.v2.createConnection({ model: 'flux-general-en' });
     * socket.on('open', () => console.log('Connected!'));
     * socket.on('message', (msg) => console.log('Transcript:', msg));
     * socket.connect(); // Actually initiates the connection
     * ```
     */
    createConnection(args) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.connect(args);
        });
    }
}
/**
 * Wrapper for Listen V2Socket that handles binary messages correctly and adds ping support.
 *
 * This wrapper ensures that both text (JSON) and binary WebSocket messages are
 * handled properly. The auto-generated socket class may not handle binary data
 * correctly, so we override the message handling to support both formats.
 * Additionally, this wrapper adds a ping() method for sending WebSocket ping
 * frames to keep connections alive (Node.js only).
 */
class WrappedListenV2Socket extends ListenV2Socket {
    constructor(args) {
        super(args);
        this.setupBinaryHandling();
    }
    setupBinaryHandling() {
        this.binaryAwareHandler = setupBinaryHandling(this.socket, this.eventHandlers);
    }
    connect() {
        // Remove duplicate listeners before calling super.connect()
        const socketAny = this;
        preventDuplicateEventListeners(this.socket, {
            handleOpen: socketAny.handleOpen,
            handleMessage: socketAny.handleMessage,
            handleClose: socketAny.handleClose,
            handleError: socketAny.handleError
        });
        resetSocketConnectionState(this.socket);
        super.connect();
        this.setupBinaryHandling();
        return this;
    }
    /**
     * Send a WebSocket ping frame to keep the connection alive.
     *
     * In Node.js, this uses the native WebSocket ping() method from the 'ws' library.
     * In browsers, WebSocket ping/pong is handled automatically by the browser and
     * cannot be manually triggered, so this method will throw an error.
     *
     * @param data Optional data to send with the ping (Node.js only)
     * @throws Error if not in Node.js environment or WebSocket is not connected
     */
    ping(data) {
        const ws = this.socket._ws;
        if (!ws) {
            throw new Error("WebSocket is not connected. Call connect() and waitForOpen() first.");
        }
        if (ws.readyState !== ws.OPEN) {
            throw new Error("WebSocket is not in OPEN state.");
        }
        // Check if we're in Node.js and the WebSocket has a ping method (from 'ws' library)
        if (RUNTIME.type === "node" && typeof ws.ping === "function") {
            // Call the native ping method from the 'ws' library
            ws.ping(data);
        }
        else {
            // In browsers, WebSocket ping/pong is automatic and not exposed to JavaScript
            throw new Error("WebSocket ping is not supported in browser environments. " +
                "Browser WebSocket connections handle ping/pong automatically. " +
                "If you need keepalive in the browser, consider sending periodic audio data or using a timer.");
        }
    }
}
/**
 * Wrapper for Speak V1Client that overrides connect to use custom websocket.
 *
 * This wrapper ensures that the connect() method uses our custom ReconnectingWebSocket
 * implementation instead of any auto-generated WebSocket handling. This guarantees
 * consistent behavior across Fern regenerations and allows us to customize
 * connection setup, authentication, and header handling.
 */
class WrappedSpeakV1Client extends SpeakV1Client {
    connect(args) {
        return __awaiter(this, void 0, void 0, function* () {
            const { headers, debug, reconnectAttempts, connectionTimeoutInSeconds, abortSignal } = args;
            const socket = yield createWebSocketConnection({
                options: this._options,
                urlPath: "/v1/speak",
                environmentKey: 'production',
                queryParams: buildQueryParams(args),
                headers,
                debug,
                reconnectAttempts,
                connectionTimeoutInSeconds,
                abortSignal,
            });
            return new WrappedSpeakV1Socket({ socket });
        });
    }
    /**
     * Creates a WebSocket connection object without actually connecting.
     * This is an alias for connect() with clearer naming - the returned socket
     * is not connected until you call socket.connect().
     *
     * Usage:
     * ```typescript
     * const socket = await client.speak.v1.createConnection({ model: 'aura-asteria-en' });
     * socket.on('open', () => console.log('Connected!'));
     * socket.on('message', (audioData) => console.log('Audio received'));
     * socket.connect(); // Actually initiates the connection
     * ```
     */
    createConnection(args) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.connect(args);
        });
    }
}
/**
 * Wrapper for Speak V1Socket that handles binary messages correctly.
 *
 * This wrapper ensures that both text (JSON) and binary WebSocket messages are
 * handled properly. The auto-generated socket class may not handle binary data
 * correctly, so we override the message handling to support both formats.
 * This is critical for audio and other binary data streaming.
 */
class WrappedSpeakV1Socket extends SpeakV1Socket {
    constructor(args) {
        super(args);
        // CRITICAL: Remove the autogenerated handleMessage that tries to parse EVERYTHING as JSON!
        // The autogenerated Socket class assumes all messages are text/JSON, but TTS sends binary audio.
        // We must remove that broken handler immediately after the parent constructor runs.
        const socketAny = this;
        if (socketAny.handleMessage) {
            this.socket.removeEventListener("message", socketAny.handleMessage);
        }
        this.setupBinaryHandling();
    }
    setupBinaryHandling() {
        this.binaryAwareHandler = setupBinaryHandling(this.socket, this.eventHandlers);
    }
    connect() {
        // Remove duplicate listeners before calling super.connect()
        const socketAny = this;
        preventDuplicateEventListeners(this.socket, {
            handleOpen: socketAny.handleOpen,
            handleMessage: socketAny.handleMessage,
            handleClose: socketAny.handleClose,
            handleError: socketAny.handleError
        });
        resetSocketConnectionState(this.socket);
        super.connect();
        this.setupBinaryHandling();
        return this;
    }
}
