// @deepgram/sdk/dist/cjs/api/types/AwsPollySpeakProvider.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AwsPollySpeakProvider = void 0;
var AwsPollySpeakProvider;
(function (AwsPollySpeakProvider) {
    /** AWS Polly voice name */
    AwsPollySpeakProvider.Voice = {
        Matthew: "Matthew",
        Joanna: "Joanna",
        Amy: "Amy",
        Emma: "Emma",
        Brian: "Brian",
        Arthur: "Arthur",
        Aria: "Aria",
        Ayanda: "Ayanda",
    };
    AwsPollySpeakProvider.Engine = {
        Generative: "generative",
        LongForm: "long-form",
        Standard: "standard",
        Neural: "neural",
    };
    let Credentials;
    (function (Credentials) {
        Credentials.Type = {
            Sts: "sts",
            Iam: "iam",
        };
    })(Credentials = AwsPollySpeakProvider.Credentials || (AwsPollySpeakProvider.Credentials = {}));
})(AwsPollySpeakProvider || (exports.AwsPollySpeakProvider = AwsPollySpeakProvider = {}));
