// @deepgram/sdk/dist/cjs/api/resources/speak/resources/v1/client/Client.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.V1Client = void 0;
const BaseClient_js_1 = require("../../../../../../BaseClient.js");
const headers_js_1 = require("../../../../../../core/headers.js");
const core = __importStar(require("../../../../../../core/index.js"));
const json_js_1 = require("../../../../../../core/json.js");
const environments = __importStar(require("../../../../../../environments.js"));
const Client_js_1 = require("../resources/audio/client/Client.js");
const Socket_js_1 = require("./Socket.js");
class V1Client {
    constructor(options = {}) {
        this._options = (0, BaseClient_js_1.normalizeClientOptionsWithAuth)(options);
    }
    get audio() {
        var _a;
        return ((_a = this._audio) !== null && _a !== void 0 ? _a : (this._audio = new Client_js_1.AudioClient(this._options)));
    }
    connect(args) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const { encoding, mip_opt_out: mipOptOut, model, sample_rate: sampleRate, queryParams, headers, debug, reconnectAttempts, connectionTimeoutInSeconds, abortSignal, } = args;
            const _queryParams = {
                encoding: encoding != null ? encoding : undefined,
                mip_opt_out: mipOptOut != null ? (typeof mipOptOut === "string" ? mipOptOut : (0, json_js_1.toJson)(mipOptOut)) : undefined,
                model: model != null ? model : undefined,
                sample_rate: sampleRate != null ? sampleRate : undefined,
            };
            const _headers = (0, headers_js_1.mergeHeaders)((0, headers_js_1.mergeOnlyDefinedHeaders)({ Authorization: args.Authorization }), headers);
            const socket = new core.ReconnectingWebSocket({
                url: core.url.join((_a = (yield core.Supplier.get(this._options.baseUrl))) !== null && _a !== void 0 ? _a : ((_b = (yield core.Supplier.get(this._options.environment))) !== null && _b !== void 0 ? _b : environments.DeepgramEnvironment.Production).production, "/v1/speak"),
                protocols: [],
                queryParameters: Object.assign(Object.assign({}, _queryParams), queryParams),
                headers: _headers,
                options: {
                    debug: debug !== null && debug !== void 0 ? debug : false,
                    maxRetries: reconnectAttempts !== null && reconnectAttempts !== void 0 ? reconnectAttempts : 30,
                    connectionTimeout: connectionTimeoutInSeconds != null ? connectionTimeoutInSeconds * 1000 : undefined,
                },
                abortSignal: abortSignal,
            });
            return new Socket_js_1.V1Socket({ socket });
        });
    }
}
exports.V1Client = V1Client;
