// @deepgram/sdk/dist/esm/middleware/index.mjs
/**
 * Deepgram SDK Middleware
 *
 * This module provides middleware for integrating Deepgram API proxying
 * into Node.js web frameworks like Express and Fastify.
 *
 * ## Features
 * - Server-side API key injection (hide keys from clients)
 * - Temporary token generation (client requests token, uses it temporarily)
 * - WebSocket proxying for Live/Agent APIs
 * - Framework integration (Raw Node.js, Express, Fastify)
 *
 * @example Express
 * ```typescript
 * import express from 'express';
 * import { createDeepgramProxy, createWebSocketHandler } from '@deepgram/sdk';
 *
 * const app = express();
 * const options = {
 *   apiKey: process.env.DEEPGRAM_API_KEY!,
 *   enableTokenAuth: true
 * };
 *
 * app.use('/api/deepgram', createDeepgramProxy(options));
 *
 * const server = app.listen(3000);
 * server.on('upgrade', createWebSocketHandler(options));
 * ```
 *
 * @example Fastify
 * ```typescript
 * import Fastify from 'fastify';
 * import { fastifyDeepgramProxy } from '@deepgram/sdk';
 *
 * const fastify = Fastify();
 *
 * fastify.register(fastifyDeepgramProxy, {
 *   prefix: '/api/deepgram',
 *   apiKey: process.env.DEEPGRAM_API_KEY!,
 *   enableTokenAuth: true
 * });
 *
 * fastify.listen({ port: 3000 });
 * ```
 *
 * @packageDocumentation
 */
// Raw Node.js HTTP/HTTPS adapter (no dependencies)
export { createHttpHandler, createUpgradeHandler as createHttpUpgradeHandler, createDeepgramHttpServer, } from "./adapters/http.mjs";
// Express adapter
export { createDeepgramProxy, createDeepgramMiddleware, createWebSocketHandler, } from "./adapters/express.mjs";
// Fastify adapter
export { fastifyDeepgramProxy, deepgramProxyPlugin, } from "./adapters/fastify.mjs";
// JWT token management
export { createProxyToken, verifyProxyToken, } from "./core/jwt.mjs";
// Core classes (for advanced usage)
export { ProxyHandler } from "./core/ProxyHandler.mjs";
export { TokenManager } from "./core/TokenManager.mjs";
export { WebSocketProxy } from "./core/WebSocketProxy.mjs";
