// @deepgram/sdk/dist/cjs/Client.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeepgramClient = void 0;
const Client_js_1 = require("./api/resources/agent/client/Client.js");
const Client_js_2 = require("./api/resources/auth/client/Client.js");
const Client_js_3 = require("./api/resources/listen/client/Client.js");
const Client_js_4 = require("./api/resources/manage/client/Client.js");
const Client_js_5 = require("./api/resources/read/client/Client.js");
const Client_js_6 = require("./api/resources/selfHosted/client/Client.js");
const Client_js_7 = require("./api/resources/speak/client/Client.js");
const BaseClient_js_1 = require("./BaseClient.js");
class DeepgramClient {
    constructor(options = {}) {
        this._options = (0, BaseClient_js_1.normalizeClientOptionsWithAuth)(options);
    }
    get agent() {
        var _a;
        return ((_a = this._agent) !== null && _a !== void 0 ? _a : (this._agent = new Client_js_1.AgentClient(this._options)));
    }
    get auth() {
        var _a;
        return ((_a = this._auth) !== null && _a !== void 0 ? _a : (this._auth = new Client_js_2.AuthClient(this._options)));
    }
    get listen() {
        var _a;
        return ((_a = this._listen) !== null && _a !== void 0 ? _a : (this._listen = new Client_js_3.ListenClient(this._options)));
    }
    get manage() {
        var _a;
        return ((_a = this._manage) !== null && _a !== void 0 ? _a : (this._manage = new Client_js_4.ManageClient(this._options)));
    }
    get read() {
        var _a;
        return ((_a = this._read) !== null && _a !== void 0 ? _a : (this._read = new Client_js_5.ReadClient(this._options)));
    }
    get selfHosted() {
        var _a;
        return ((_a = this._selfHosted) !== null && _a !== void 0 ? _a : (this._selfHosted = new Client_js_6.SelfHostedClient(this._options)));
    }
    get speak() {
        var _a;
        return ((_a = this._speak) !== null && _a !== void 0 ? _a : (this._speak = new Client_js_7.SpeakClient(this._options)));
    }
}
exports.DeepgramClient = DeepgramClient;
