// @deepgram/sdk/dist/esm/middleware/core/TokenManager.mjs
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { CustomDeepgramClient as DeepgramClient } from "../../CustomClient.mjs";
import { createProxyToken } from "./jwt.mjs";
/**
 * Manages temporary token generation for client authentication
 */
export class TokenManager {
    constructor(options) {
        var _a, _b;
        this.client = new DeepgramClient({ apiKey: options.apiKey });
        this.apiKey = options.apiKey;
        this.tokenMode = (_a = options.tokenMode) !== null && _a !== void 0 ? _a : "deepgram";
        this.defaultExpiration = (_b = options.defaultTokenExpiration) !== null && _b !== void 0 ? _b : 3600;
    }
    /**
     * Generate a temporary access token
     * Uses either Deepgram's native token API or JWT signing based on tokenMode
     */
    generateToken(ttlSeconds) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.tokenMode === "jwt") {
                return createProxyToken(this.apiKey, { ttl: ttlSeconds });
            }
            const response = yield this.client.auth.v1.tokens.grant({
                ttl_seconds: ttlSeconds,
            });
            return response.access_token;
        });
    }
    /**
     * Handle token generation request from client
     */
    handleTokenRequest(req, res) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            try {
                // Parse request body
                const body = yield this.parseBody(req);
                const tokenRequest = body;
                // Generate token with requested or default TTL
                const ttlSeconds = (_a = tokenRequest.ttlSeconds) !== null && _a !== void 0 ? _a : this.defaultExpiration;
                const token = yield this.generateToken(ttlSeconds);
                // Send response
                const response = {
                    token,
                    expiresIn: ttlSeconds,
                };
                res.writeHead(200, {
                    "Content-Type": "application/json",
                    "Access-Control-Allow-Origin": "*",
                    "Access-Control-Allow-Methods": "POST, OPTIONS",
                    "Access-Control-Allow-Headers": "Content-Type",
                });
                res.end(JSON.stringify(response));
            }
            catch (error) {
                this.handleError(error, res);
            }
        });
    }
    /**
     * Parse JSON request body
     */
    parseBody(req) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                let body = "";
                req.on("data", (chunk) => {
                    body += chunk.toString();
                });
                req.on("end", () => {
                    try {
                        resolve(body ? JSON.parse(body) : {});
                    }
                    catch (error) {
                        reject(new Error("Invalid JSON in request body"));
                    }
                });
                req.on("error", reject);
            });
        });
    }
    /**
     * Handle errors and send appropriate response
     */
    handleError(error, res) {
        console.error("TokenManager error:", error);
        const message = error instanceof Error ? error.message : "Internal server error";
        const statusCode = message.includes("Invalid JSON") ? 400 : 500;
        res.writeHead(statusCode, {
            "Content-Type": "application/json",
            "Access-Control-Allow-Origin": "*",
        });
        res.end(JSON.stringify({
            error: message,
        }));
    }
}
