// @deepgram/sdk/dist/esm/middleware/core/WebSocketProxy.mjs
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { URL } from "node:url";
import WebSocket from "ws";
import { verifyProxyToken } from "./jwt.mjs";
/**
 * Handles WebSocket proxying to Deepgram API
 */
export class WebSocketProxy {
    constructor(options) {
        var _a, _b;
        this.apiKey = options.apiKey;
        this.tokenMode = (_a = options.tokenMode) !== null && _a !== void 0 ? _a : "deepgram";
        // Convert HTTP(S) base URL to WebSocket URL
        this.baseUrl = ((_b = options.baseUrl) !== null && _b !== void 0 ? _b : "https://api.deepgram.com")
            .replace(/^https:/, "wss:")
            .replace(/^http:/, "ws:");
    }
    /**
     * Handle WebSocket upgrade request
     */
    handleUpgrade(req, socket, head) {
        this.handleUpgradeAsync(req, socket, head).catch((error) => {
            console.error("WebSocket upgrade error:", error);
            socket.destroy();
        });
    }
    /**
     * Async version of handleUpgrade to support JWT verification
     */
    handleUpgradeAsync(req, socket, head) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                // Build Deepgram WebSocket URL
                const targetUrl = this.buildWebSocketUrl(req);
                // Prepare headers (may verify JWT)
                const headers = yield this.prepareHeaders(req);
                // Connect to Deepgram WebSocket
                const deepgramSocket = new WebSocket(targetUrl, { headers });
                // Handle connection open
                deepgramSocket.on("open", () => {
                    // Complete the client upgrade
                    socket.write("HTTP/1.1 101 Switching Protocols\r\n" +
                        "Upgrade: websocket\r\n" +
                        "Connection: Upgrade\r\n" +
                        "\r\n");
                    // Create client WebSocket
                    const clientSocket = new WebSocket(null, null);
                    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- WebSocket internals
                    clientSocket.setSocket(socket, head, { maxPayload: 100 * 1024 * 1024 });
                    // Forward messages bidirectionally
                    this.forwardMessages(clientSocket, deepgramSocket);
                });
                // Handle connection errors
                deepgramSocket.on("error", (error) => {
                    console.error("WebSocket proxy error (Deepgram):", error);
                    socket.destroy();
                });
            }
            catch (error) {
                console.error("WebSocket upgrade error:", error);
                throw error;
            }
        });
    }
    /**
     * Build WebSocket URL for Deepgram
     */
    buildWebSocketUrl(req) {
        var _a;
        // Strip /api/deepgram prefix if present
        const cleanUrl = ((_a = req.url) !== null && _a !== void 0 ? _a : "/").replace(/^\/api\/deepgram/, "");
        const url = new URL(cleanUrl, this.baseUrl);
        return url.toString();
    }
    /**
     * Prepare headers for Deepgram WebSocket connection
     */
    prepareHeaders(req) {
        return __awaiter(this, void 0, void 0, function* () {
            const headers = {};
            // Copy relevant headers
            const allowedHeaders = [
                "user-agent",
                "x-fern-language",
                "x-fern-sdk-name",
                "x-fern-sdk-version",
                "x-deepgram-session-id",
            ];
            for (const header of allowedHeaders) {
                const value = req.headers[header];
                if (value) {
                    headers[header] = Array.isArray(value) ? value[0] : value;
                }
            }
            // Authentication logic (similar to ProxyHandler)
            const clientAuth = req.headers.authorization;
            if (clientAuth && this.tokenMode === "jwt" && clientAuth.startsWith("Bearer ")) {
                const token = clientAuth.replace("Bearer ", "");
                console.log(`[WebSocketProxy] Client provided JWT token, verifying...`);
                try {
                    yield verifyProxyToken(token, this.apiKey);
                    console.log(`[WebSocketProxy] JWT verified successfully, using server API key`);
                    headers.authorization = `Token ${this.apiKey}`;
                }
                catch (error) {
                    console.error(`[WebSocketProxy] JWT verification failed:`, error);
                    throw new Error("Invalid or expired token");
                }
            }
            else {
                // Use server API key
                headers.authorization = `Token ${this.apiKey}`;
            }
            return headers;
        });
    }
    /**
     * Forward WebSocket messages bidirectionally
     */
    forwardMessages(clientSocket, deepgramSocket) {
        // Forward client messages to Deepgram
        clientSocket.on("message", (data, isBinary) => {
            if (deepgramSocket.readyState === WebSocket.OPEN) {
                deepgramSocket.send(data, { binary: isBinary });
            }
        });
        // Forward Deepgram messages to client
        deepgramSocket.on("message", (data, isBinary) => {
            if (clientSocket.readyState === WebSocket.OPEN) {
                clientSocket.send(data, { binary: isBinary });
            }
        });
        // Handle closures
        clientSocket.on("close", (code, reason) => {
            if (deepgramSocket.readyState === WebSocket.OPEN) {
                deepgramSocket.close(code, reason);
            }
        });
        deepgramSocket.on("close", (code, reason) => {
            if (clientSocket.readyState === WebSocket.OPEN) {
                clientSocket.close(code, reason);
            }
        });
        // Handle errors
        clientSocket.on("error", (error) => {
            console.error("WebSocket error (client):", error);
            deepgramSocket.close();
        });
        deepgramSocket.on("error", (error) => {
            console.error("WebSocket error (Deepgram):", error);
            clientSocket.close();
        });
    }
}
