// @deepgram/sdk/dist/cjs/middleware/index.js
"use strict";
/**
 * Deepgram SDK Middleware
 *
 * This module provides middleware for integrating Deepgram API proxying
 * into Node.js web frameworks like Express and Fastify.
 *
 * ## Features
 * - Server-side API key injection (hide keys from clients)
 * - Temporary token generation (client requests token, uses it temporarily)
 * - WebSocket proxying for Live/Agent APIs
 * - Framework integration (Raw Node.js, Express, Fastify)
 *
 * @example Express
 * ```typescript
 * import express from 'express';
 * import { createDeepgramProxy, createWebSocketHandler } from '@deepgram/sdk';
 *
 * const app = express();
 * const options = {
 *   apiKey: process.env.DEEPGRAM_API_KEY!,
 *   enableTokenAuth: true
 * };
 *
 * app.use('/api/deepgram', createDeepgramProxy(options));
 *
 * const server = app.listen(3000);
 * server.on('upgrade', createWebSocketHandler(options));
 * ```
 *
 * @example Fastify
 * ```typescript
 * import Fastify from 'fastify';
 * import { fastifyDeepgramProxy } from '@deepgram/sdk';
 *
 * const fastify = Fastify();
 *
 * fastify.register(fastifyDeepgramProxy, {
 *   prefix: '/api/deepgram',
 *   apiKey: process.env.DEEPGRAM_API_KEY!,
 *   enableTokenAuth: true
 * });
 *
 * fastify.listen({ port: 3000 });
 * ```
 *
 * @packageDocumentation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketProxy = exports.TokenManager = exports.ProxyHandler = exports.verifyProxyToken = exports.createProxyToken = exports.deepgramProxyPlugin = exports.fastifyDeepgramProxy = exports.createWebSocketHandler = exports.createDeepgramMiddleware = exports.createDeepgramProxy = exports.createDeepgramHttpServer = exports.createHttpUpgradeHandler = exports.createHttpHandler = void 0;
// Raw Node.js HTTP/HTTPS adapter (no dependencies)
var http_js_1 = require("./adapters/http.js");
Object.defineProperty(exports, "createHttpHandler", { enumerable: true, get: function () { return http_js_1.createHttpHandler; } });
Object.defineProperty(exports, "createHttpUpgradeHandler", { enumerable: true, get: function () { return http_js_1.createUpgradeHandler; } });
Object.defineProperty(exports, "createDeepgramHttpServer", { enumerable: true, get: function () { return http_js_1.createDeepgramHttpServer; } });
// Express adapter
var express_js_1 = require("./adapters/express.js");
Object.defineProperty(exports, "createDeepgramProxy", { enumerable: true, get: function () { return express_js_1.createDeepgramProxy; } });
Object.defineProperty(exports, "createDeepgramMiddleware", { enumerable: true, get: function () { return express_js_1.createDeepgramMiddleware; } });
Object.defineProperty(exports, "createWebSocketHandler", { enumerable: true, get: function () { return express_js_1.createWebSocketHandler; } });
// Fastify adapter
var fastify_js_1 = require("./adapters/fastify.js");
Object.defineProperty(exports, "fastifyDeepgramProxy", { enumerable: true, get: function () { return fastify_js_1.fastifyDeepgramProxy; } });
Object.defineProperty(exports, "deepgramProxyPlugin", { enumerable: true, get: function () { return fastify_js_1.deepgramProxyPlugin; } });
// JWT token management
var jwt_js_1 = require("./core/jwt.js");
Object.defineProperty(exports, "createProxyToken", { enumerable: true, get: function () { return jwt_js_1.createProxyToken; } });
Object.defineProperty(exports, "verifyProxyToken", { enumerable: true, get: function () { return jwt_js_1.verifyProxyToken; } });
// Core classes (for advanced usage)
var ProxyHandler_js_1 = require("./core/ProxyHandler.js");
Object.defineProperty(exports, "ProxyHandler", { enumerable: true, get: function () { return ProxyHandler_js_1.ProxyHandler; } });
var TokenManager_js_1 = require("./core/TokenManager.js");
Object.defineProperty(exports, "TokenManager", { enumerable: true, get: function () { return TokenManager_js_1.TokenManager; } });
var WebSocketProxy_js_1 = require("./core/WebSocketProxy.js");
Object.defineProperty(exports, "WebSocketProxy", { enumerable: true, get: function () { return WebSocketProxy_js_1.WebSocketProxy; } });
