// @deepgram/sdk/dist/cjs/api/resources/listen/resources/v1/resources/media/client/Client.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaClient = void 0;
const BaseClient_js_1 = require("../../../../../../../../BaseClient.js");
const headers_js_1 = require("../../../../../../../../core/headers.js");
const core = __importStar(require("../../../../../../../../core/index.js"));
const environments = __importStar(require("../../../../../../../../environments.js"));
const handleNonStatusCodeError_js_1 = require("../../../../../../../../errors/handleNonStatusCodeError.js");
const errors = __importStar(require("../../../../../../../../errors/index.js"));
const Deepgram = __importStar(require("../../../../../../../index.js"));
class MediaClient {
    constructor(options = {}) {
        this._options = (0, BaseClient_js_1.normalizeClientOptionsWithAuth)(options);
    }
    /**
     * Transcribe audio and video using Deepgram's speech-to-text REST API
     *
     * @param {Deepgram.listen.v1.ListenV1RequestUrl} request
     * @param {MediaClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link Deepgram.BadRequestError}
     *
     * @example
     *     await client.listen.v1.media.transcribeUrl({
     *         callback: "callback",
     *         callback_method: "POST",
     *         extra: "extra",
     *         sentiment: true,
     *         summarize: "v2",
     *         tag: "tag",
     *         topics: true,
     *         custom_topic: "custom_topic",
     *         custom_topic_mode: "extended",
     *         intents: true,
     *         custom_intent: "custom_intent",
     *         custom_intent_mode: "extended",
     *         detect_entities: true,
     *         detect_language: true,
     *         diarize: true,
     *         dictation: true,
     *         encoding: "linear16",
     *         filler_words: true,
     *         keywords: "keywords",
     *         language: "language",
     *         measurements: true,
     *         model: "nova-3",
     *         multichannel: true,
     *         numerals: true,
     *         paragraphs: true,
     *         profanity_filter: true,
     *         punctuate: true,
     *         redact: "redact",
     *         replace: "replace",
     *         search: "search",
     *         smart_format: true,
     *         utterances: true,
     *         utt_split: 1.1,
     *         version: "latest",
     *         mip_opt_out: true,
     *         url: "https://dpgr.am/spacewalk.wav"
     *     })
     */
    transcribeUrl(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__transcribeUrl(request, requestOptions));
    }
    __transcribeUrl(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
            const { callback, callback_method: callbackMethod, extra, sentiment, summarize, tag, topics, custom_topic: customTopic, custom_topic_mode: customTopicMode, intents, custom_intent: customIntent, custom_intent_mode: customIntentMode, detect_entities: detectEntities, detect_language: detectLanguage, diarize, dictation, encoding, filler_words: fillerWords, keyterm, keywords, language, measurements, model, multichannel, numerals, paragraphs, profanity_filter: profanityFilter, punctuate, redact, replace, search, smart_format: smartFormat, utterances, utt_split: uttSplit, version, mip_opt_out: mipOptOut } = request, _body = __rest(request, ["callback", "callback_method", "extra", "sentiment", "summarize", "tag", "topics", "custom_topic", "custom_topic_mode", "intents", "custom_intent", "custom_intent_mode", "detect_entities", "detect_language", "diarize", "dictation", "encoding", "filler_words", "keyterm", "keywords", "language", "measurements", "model", "multichannel", "numerals", "paragraphs", "profanity_filter", "punctuate", "redact", "replace", "search", "smart_format", "utterances", "utt_split", "version", "mip_opt_out"]);
            const _queryParams = {
                callback,
                callback_method: callbackMethod != null ? callbackMethod : undefined,
                extra,
                sentiment,
                summarize: summarize != null ? summarize : undefined,
                tag,
                topics,
                custom_topic: customTopic,
                custom_topic_mode: customTopicMode != null ? customTopicMode : undefined,
                intents,
                custom_intent: customIntent,
                custom_intent_mode: customIntentMode != null ? customIntentMode : undefined,
                detect_entities: detectEntities,
                detect_language: detectLanguage,
                diarize,
                dictation,
                encoding: encoding != null ? encoding : undefined,
                filler_words: fillerWords,
                keyterm,
                keywords,
                language,
                measurements,
                model: model != null ? model : undefined,
                multichannel,
                numerals,
                paragraphs,
                profanity_filter: profanityFilter,
                punctuate,
                redact,
                replace,
                search,
                smart_format: smartFormat,
                utterances,
                utt_split: uttSplit,
                version: version != null ? version : undefined,
                mip_opt_out: mipOptOut,
            };
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = (0, headers_js_1.mergeHeaders)(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield ((_b = this._options.fetcher) !== null && _b !== void 0 ? _b : core.fetcher)({
                url: core.url.join((_c = (yield core.Supplier.get(this._options.baseUrl))) !== null && _c !== void 0 ? _c : ((_d = (yield core.Supplier.get(this._options.environment))) !== null && _d !== void 0 ? _d : environments.DeepgramEnvironment.Production).base, "v1/listen"),
                method: "POST",
                headers: _headers,
                contentType: "application/json",
                queryParameters: Object.assign(Object.assign({}, _queryParams), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams),
                requestType: "json",
                body: _body,
                timeoutMs: ((_g = (_e = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _e !== void 0 ? _e : (_f = this._options) === null || _f === void 0 ? void 0 : _f.timeoutInSeconds) !== null && _g !== void 0 ? _g : 60) * 1000,
                maxRetries: (_h = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _h !== void 0 ? _h : (_j = this._options) === null || _j === void 0 ? void 0 : _j.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_k = this._options) === null || _k === void 0 ? void 0 : _k.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return {
                    data: _response.body,
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new Deepgram.BadRequestError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.DeepgramError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return (0, handleNonStatusCodeError_js_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "POST", "/v1/listen");
        });
    }
    /**
     * Transcribe audio and video using Deepgram's speech-to-text REST API
     *
     * @param {core.file.Uploadable} uploadable
     * @param {Deepgram.listen.v1.MediaTranscribeRequestOctetStream} request
     * @param {MediaClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link Deepgram.BadRequestError}
     *
     * @example
     *     import { createReadStream } from "fs";
     *     await client.listen.v1.media.transcribeFile(createReadStream("path/to/file"), {})
     */
    transcribeFile(uploadable, request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__transcribeFile(uploadable, request, requestOptions));
    }
    __transcribeFile(uploadable, request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
            const _queryParams = {
                callback: request.callback,
                callback_method: request.callback_method != null ? request.callback_method : undefined,
                extra: request.extra,
                sentiment: request.sentiment,
                summarize: request.summarize != null ? request.summarize : undefined,
                tag: request.tag,
                topics: request.topics,
                custom_topic: request.custom_topic,
                custom_topic_mode: request.custom_topic_mode != null ? request.custom_topic_mode : undefined,
                intents: request.intents,
                custom_intent: request.custom_intent,
                custom_intent_mode: request.custom_intent_mode != null ? request.custom_intent_mode : undefined,
                detect_entities: request.detect_entities,
                detect_language: request.detect_language,
                diarize: request.diarize,
                dictation: request.dictation,
                encoding: request.encoding != null ? request.encoding : undefined,
                filler_words: request.filler_words,
                keyterm: request.keyterm,
                keywords: request.keywords,
                language: request.language,
                measurements: request.measurements,
                model: request.model != null ? request.model : undefined,
                multichannel: request.multichannel,
                numerals: request.numerals,
                paragraphs: request.paragraphs,
                profanity_filter: request.profanity_filter,
                punctuate: request.punctuate,
                redact: request.redact,
                replace: request.replace,
                search: request.search,
                smart_format: request.smart_format,
                utterances: request.utterances,
                utt_split: request.utt_split,
                version: request.version != null ? request.version : undefined,
                mip_opt_out: request.mip_opt_out,
            };
            const _binaryUploadRequest = yield core.file.toBinaryUploadRequest(uploadable);
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = (0, headers_js_1.mergeHeaders)(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, _binaryUploadRequest.headers, requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield ((_b = this._options.fetcher) !== null && _b !== void 0 ? _b : core.fetcher)({
                url: core.url.join((_c = (yield core.Supplier.get(this._options.baseUrl))) !== null && _c !== void 0 ? _c : ((_d = (yield core.Supplier.get(this._options.environment))) !== null && _d !== void 0 ? _d : environments.DeepgramEnvironment.Production).base, "v1/listen"),
                method: "POST",
                headers: _headers,
                contentType: "application/octet-stream",
                queryParameters: Object.assign(Object.assign({}, _queryParams), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams),
                requestType: "bytes",
                duplex: "half",
                body: _binaryUploadRequest.body,
                timeoutMs: ((_g = (_e = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _e !== void 0 ? _e : (_f = this._options) === null || _f === void 0 ? void 0 : _f.timeoutInSeconds) !== null && _g !== void 0 ? _g : 60) * 1000,
                maxRetries: (_h = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _h !== void 0 ? _h : (_j = this._options) === null || _j === void 0 ? void 0 : _j.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_k = this._options) === null || _k === void 0 ? void 0 : _k.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return {
                    data: _response.body,
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new Deepgram.BadRequestError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.DeepgramError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return (0, handleNonStatusCodeError_js_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "POST", "/v1/listen");
        });
    }
}
exports.MediaClient = MediaClient;
