// @deepgram/sdk/dist/esm/api/resources/manage/resources/v1/resources/projects/resources/usage/resources/breakdown/client/Client.mjs
// This file was auto-generated by Fern from our API Definition.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { normalizeClientOptionsWithAuth, } from "../../../../../../../../../../../../BaseClient.mjs";
import { mergeHeaders } from "../../../../../../../../../../../../core/headers.mjs";
import * as core from "../../../../../../../../../../../../core/index.mjs";
import * as environments from "../../../../../../../../../../../../environments.mjs";
import { handleNonStatusCodeError } from "../../../../../../../../../../../../errors/handleNonStatusCodeError.mjs";
import * as errors from "../../../../../../../../../../../../errors/index.mjs";
import * as Deepgram from "../../../../../../../../../../../index.mjs";
export class BreakdownClient {
    constructor(options = {}) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    /**
     * Retrieves the usage breakdown for a specific project, with various filter options by API feature or by groupings. Setting a feature (e.g. diarize) to true includes requests that used that feature, while false excludes requests that used it. Multiple true filters are combined with OR logic, while false filters use AND logic.
     *
     * @param {string} project_id - The unique identifier of the project
     * @param {Deepgram.manage.v1.projects.usage.BreakdownGetRequest} request
     * @param {BreakdownClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link Deepgram.BadRequestError}
     *
     * @example
     *     await client.manage.v1.projects.usage.breakdown.get("123456-7890-1234-5678-901234", {
     *         start: "start",
     *         end: "end",
     *         grouping: "accessor",
     *         accessor: "12345678-1234-1234-1234-123456789012",
     *         alternatives: true,
     *         callback_method: true,
     *         callback: true,
     *         channels: true,
     *         custom_intent_mode: true,
     *         custom_intent: true,
     *         custom_topic_mode: true,
     *         custom_topic: true,
     *         deployment: "hosted",
     *         detect_entities: true,
     *         detect_language: true,
     *         diarize: true,
     *         dictation: true,
     *         encoding: true,
     *         endpoint: "listen",
     *         extra: true,
     *         filler_words: true,
     *         intents: true,
     *         keyterm: true,
     *         keywords: true,
     *         language: true,
     *         measurements: true,
     *         method: "sync",
     *         model: "6f548761-c9c0-429a-9315-11a1d28499c8",
     *         multichannel: true,
     *         numerals: true,
     *         paragraphs: true,
     *         profanity_filter: true,
     *         punctuate: true,
     *         redact: true,
     *         replace: true,
     *         sample_rate: true,
     *         search: true,
     *         sentiment: true,
     *         smart_format: true,
     *         summarize: true,
     *         tag: "tag1",
     *         topics: true,
     *         utt_split: true,
     *         utterances: true,
     *         version: true
     *     })
     */
    get(project_id, request = {}, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__get(project_id, request, requestOptions));
    }
    __get(project_id_1) {
        return __awaiter(this, arguments, void 0, function* (project_id, request = {}, requestOptions) {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
            const { start, end, grouping, accessor, alternatives, callback_method: callbackMethod, callback, channels, custom_intent_mode: customIntentMode, custom_intent: customIntent, custom_topic_mode: customTopicMode, custom_topic: customTopic, deployment, detect_entities: detectEntities, detect_language: detectLanguage, diarize, dictation, encoding, endpoint, extra, filler_words: fillerWords, intents, keyterm, keywords, language, measurements, method, model, multichannel, numerals, paragraphs, profanity_filter: profanityFilter, punctuate, redact, replace, sample_rate: sampleRate, search, sentiment, smart_format: smartFormat, summarize, tag, topics, utt_split: uttSplit, utterances, version, } = request;
            const _queryParams = {
                start,
                end,
                grouping: grouping != null ? grouping : undefined,
                accessor,
                alternatives,
                callback_method: callbackMethod,
                callback,
                channels,
                custom_intent_mode: customIntentMode,
                custom_intent: customIntent,
                custom_topic_mode: customTopicMode,
                custom_topic: customTopic,
                deployment: deployment != null ? deployment : undefined,
                detect_entities: detectEntities,
                detect_language: detectLanguage,
                diarize,
                dictation,
                encoding,
                endpoint: endpoint != null ? endpoint : undefined,
                extra,
                filler_words: fillerWords,
                intents,
                keyterm,
                keywords,
                language,
                measurements,
                method: method != null ? method : undefined,
                model,
                multichannel,
                numerals,
                paragraphs,
                profanity_filter: profanityFilter,
                punctuate,
                redact,
                replace,
                sample_rate: sampleRate,
                search,
                sentiment,
                smart_format: smartFormat,
                summarize,
                tag,
                topics,
                utt_split: uttSplit,
                utterances,
                version,
            };
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = mergeHeaders(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield ((_b = this._options.fetcher) !== null && _b !== void 0 ? _b : core.fetcher)({
                url: core.url.join((_c = (yield core.Supplier.get(this._options.baseUrl))) !== null && _c !== void 0 ? _c : ((_d = (yield core.Supplier.get(this._options.environment))) !== null && _d !== void 0 ? _d : environments.DeepgramEnvironment.Production).base, `v1/projects/${core.url.encodePathParam(project_id)}/usage/breakdown`),
                method: "GET",
                headers: _headers,
                queryParameters: Object.assign(Object.assign({}, _queryParams), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams),
                timeoutMs: ((_g = (_e = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _e !== void 0 ? _e : (_f = this._options) === null || _f === void 0 ? void 0 : _f.timeoutInSeconds) !== null && _g !== void 0 ? _g : 60) * 1000,
                maxRetries: (_h = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _h !== void 0 ? _h : (_j = this._options) === null || _j === void 0 ? void 0 : _j.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_k = this._options) === null || _k === void 0 ? void 0 : _k.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return { data: _response.body, rawResponse: _response.rawResponse };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new Deepgram.BadRequestError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.DeepgramError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/v1/projects/{project_id}/usage/breakdown");
        });
    }
}
