// @deepgram/sdk/dist/cjs/api/resources/listen/resources/v2/types/ListenV2TurnInfo.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListenV2TurnInfo = void 0;
var ListenV2TurnInfo;
(function (ListenV2TurnInfo) {
    /**
     * The type of event being reported.
     *
     * - **Update** - Additional audio has been transcribed, but the turn state hasn't changed
     * - **StartOfTurn** - The user has begun speaking for the first time in the turn
     * - **EagerEndOfTurn** - The system has moderate confidence that the user has finished speaking for the turn. This is an opportunity to begin preparing an agent reply
     * - **TurnResumed** - The system detected that speech had ended and therefore sent an **EagerEndOfTurn** event, but speech is actually continuing for this turn
     * - **EndOfTurn** - The user has finished speaking for the turn
     */
    ListenV2TurnInfo.Event = {
        Update: "Update",
        StartOfTurn: "StartOfTurn",
        EagerEndOfTurn: "EagerEndOfTurn",
        TurnResumed: "TurnResumed",
        EndOfTurn: "EndOfTurn",
    };
})(ListenV2TurnInfo || (exports.ListenV2TurnInfo = ListenV2TurnInfo = {}));
